src/nvim/winfloat.c

```diff
@@ -433,9 +594,148 @@ win_T *win_float_create(bool enter, bool new_buf)
+bool win_float_close(bool preview)
+{
+  win_T *wp = win_float_find_preview(preview);
+  return wp && win_close(wp, false, false) != FAIL;
+}
+
+/// Set bufname as title for a floating window.
+/// Title position is center.
+///
+/// @param wp A pointer of win_T
+/// @param redraw bool
+/// @return
+void win_float_set_title(win_T *wp)
+{
+  if (!wp->w_floating || !wp->w_config.border) {
+    return;
+  }
+
+  if (wp->w_config.title) {
+    clear_virttext(&wp->w_config.title_chunks);
+  }
+  wp->w_config.title = true;
+  wp->w_config.title_pos = kAlignCenter;
+  wp->w_config.title_width = (int)mb_string2cells(wp->w_buffer->b_fname);
+  kv_push(*(&wp->w_config.title_chunks), ((VirtTextChunk){
+    .text = xstrdup(wp->w_buffer->b_fname), .hl_id = -1
+  }));
+  win_config_float(wp, wp->w_config);
+}
+
+/// adjust a preview floating window postion to fit screen and buffer in wp.
+///
+/// @param wp A pointer of win_T
+void win_float_adjust_position(win_T *wp)
+{
+  if (!wp->w_floating) {
+    return;
+  }
+
+  // Get cursor screen coordinates
+  int cursor_row = curwin->w_winrow + curwin->w_wrow;
+  int cursor_col = curwin->w_wincol + curwin->w_wcol;
+
+  int border_extra = wp->w_config.border ? 1 : 0;  // Usually border is 1, not 2
+
+  // Calculate available space in four directions
+  int right_extra = Columns - cursor_col - border_extra;
+  int left_extra = cursor_col - border_extra;
+  int below_extra = Rows - cursor_row - 1 - border_extra;  // -1 for cmdline
+  int above_extra = cursor_row - border_extra;
+
+  // Determine horizontal position
+  bool place_right = false;
+  int available_width = wp->w_config.width;
+
+  if (available_width <= right_extra) {
+    place_right = true;  // Prefer right side
+  } else if (available_width <= left_extra) {
+    place_right = false;  // Place on left side
+  } else {
+    // Neither side has enough space, choose the larger one
+    if (right_extra >= left_extra) {
+      place_right = true;
+      wp->w_config.width = right_extra;
+    } else {
+      place_right = false;
+      wp->w_config.width = left_extra;
+    }
+  }
+
+  // Determine vertical position
+  bool place_below = true;
+  int available_height = wp->w_config.height;
+
+  if (available_height <= below_extra) {
+    place_below = true;  // Prefer below cursor
+  } else if (available_height <= above_extra) {
+    place_below = false;  // Place above cursor
+  } else {
+    // Neither direction has enough space, choose the larger one
+    if (below_extra >= above_extra) {
+      place_below = true;
+      wp->w_config.height = below_extra;
+    } else {
+      place_below = false;
+      wp->w_config.height = above_extra;
+    }
+  }
+
+  // Set anchor and position
+  wp->w_config.anchor = 0;  // Reset anchor
+
+  if (place_below && place_right) {
+    // Bottom-right: anchor NW, position at cursor bottom-right
+    wp->w_config.anchor = 0;  // NW
+    wp->w_config.row = cursor_row + 1;
+    wp->w_config.col = cursor_col + 1;
```

Comment:
don't we already have placement code somewhere?

Reply:
popupmenu have a smimilar code. but logic need extra handle.